src/number_theory/class_number/admissible_absolute_value.lean

```diff
@@ -0,0 +1,418 @@
+  { intro i,
+    exact floor_nonneg.mpr (div_nonneg (cast_nonneg.mpr (mod_nonneg _ hb)) hbε.le) },
+  refine ⟨λ i, ⟨nat_abs (floor ((A i % b : ℤ) / (abs b • ε) : ℝ)), _⟩, _⟩,
+  { rw [← coe_nat_lt, nat_abs_of_nonneg (hfloor i), floor_lt],
+    apply lt_of_lt_of_le _ (le_nat_ceil _),
+    rw [algebra.smul_def, ring_hom.eq_int_cast, ← div_div_eq_div_mul, div_lt_div_right hε,
+        div_lt_iff hb', one_mul, cast_lt],
+    exact int.mod_lt _ hb },
+  intros i₀ i₁ hi,
+  have hi : (⌊↑(A i₀ % b) / abs b • ε⌋.nat_abs : ℤ) = ⌊↑(A i₁ % b) / abs b • ε⌋.nat_abs :=
+    congr_arg (coe : ℕ → ℤ) (subtype.mk_eq_mk.mp hi),
+  rw [nat_abs_of_nonneg (hfloor i₀), nat_abs_of_nonneg (hfloor i₁)] at hi,
+  have hi := abs_sub_lt_one_of_floor_eq_floor hi,
+  rw [abs_sub_comm, ← sub_div, abs_div, abs_of_nonneg hbε.le, div_lt_iff hbε, one_mul] at hi,
+  rwa [int.cast_abs, int.cast_sub]
+end
+
+/-- `abs : ℤ → ℤ` is an admissible absolute value -/
+noncomputable def abs_is_admissible : is_admissible absolute_value.abs :=
+{ card := λ ε, nat_ceil (1 / ε),
+  exists_partition' := λ n ε hε b hb, exists_partition_int n hε hb,
+  .. absolute_value.abs_is_euclidean }
+
+noncomputable instance : inhabited (is_admissible absolute_value.abs) :=
+⟨abs_is_admissible⟩
+
+end int
+
+end absolute_value
+
+namespace polynomial
+
+open absolute_value real
+
+variables {Fq : Type*} [field Fq] [fintype Fq]
+
+/-- If `A` is a family of enough low-degree polynomials over a finite field, there is a
+pair of equal elements in `A`. -/
+lemma exists_eq_polynomial {d : ℕ} {m : ℕ} (hm : fintype.card Fq ^ d ≤ m) (b : polynomial Fq)
+  (hb : nat_degree b ≤ d) (A : fin m.succ → polynomial Fq) (hA : ∀ i, degree (A i) < degree b) :
+  ∃ i₀ i₁, i₀ ≠ i₁ ∧ A i₁ = A i₀ :=
+begin
+  -- Since there are > q^d elements of A, and only q^d choices for the highest `d` coefficients,
+  -- there must be two elements of A with the same coefficients at
+  -- `0`, ... `degree b - 1` ≤ `d - 1`.
+  -- In other words, the following map is not injective:
+  set f : fin m.succ → (fin d → Fq) := λ i j, (A i).coeff j,
+  have : fintype.card (fin d → Fq) < fintype.card (fin m.succ),
+  { simpa using lt_of_le_of_lt hm (nat.lt_succ_self m) },
+  -- Therefore, the differences have all coefficients higher than `deg b - d` equal.
+  obtain ⟨i₀, i₁, i_ne, i_eq⟩ := fintype.exists_ne_map_eq_of_card_lt f this,
+  use [i₀, i₁, i_ne],
+  ext j,
+  -- The coefficients higher than `deg b` are the same because they are equal to 0.
+  by_cases hbj : degree b ≤ j,
+  { rw [coeff_eq_zero_of_degree_lt (lt_of_lt_of_le (hA _) hbj),
+        coeff_eq_zero_of_degree_lt (lt_of_lt_of_le (hA _) hbj)] },
+  -- So we only need to look for the coefficients between `0` and `deg b`.
+  rw not_le at hbj,
+  apply congr_fun i_eq.symm ⟨j, _⟩,
+  exact lt_of_lt_of_le (coe_lt_degree.mp hbj) hb
+end
+
+/-- If `A` is a family of enough low-degree polynomials over a finite field,
+there is a pair of elements in `A` (not necessarily distinct), such that their
```

Comment:
I have tried to clarify the docstring. The point is: the indices `i₀, i₁` are different (otherwise everything becomes trivial), but `A` doesn't have to be injective, so `A i₀ = A i₁` is still allowed.